File: frontend/src/modules/inventario/services/inventario.service.ts
import { api } from '@/shared/services/api'
import type {
  Inventario,
  InventarioResumenConProducto,
  MovimientoInventario,
  MovimientoInventarioCreate,
} from '../types/inventario.types'

export interface MovimientoEntradaReporte {
  producto_id: number
  producto_referencia: string
  producto_material: string
  cantidad_total: number
}

export interface MovimientoEntradaDetalle {
  id: number
  producto_id: number
  producto_referencia: string
  producto_material: string
  cantidad: number
  created_at: string
}

export interface MovimientoInventarioReporteDetalle {
  id: number
  producto_id: number
  producto_referencia: string
  producto_material: string
  tipo: 'entrada' | 'salida' | 'ajuste'
  cantidad: number
  stock_anterior: number
  stock_nuevo: number
  referencia_tipo?: string | null
  referencia_id?: number | null
  pedido_id?: number | null
  numero_pedido?: string | null
  cliente_razon_social?: string | null
  numero_remision?: string | null
  motivo?: string | null
  usuario_nombre?: string | null
  created_at: string
}

export const inventarioService = {
  list: (params?: { skip?: number; limit?: number; pedido_ids?: string }) =>
    api.get<InventarioResumenConProducto[]>('/inventario', { params }).then((r) => r.data),

  getByProducto: (productoId: number) =>
    api.get<Inventario>(`/inventario/producto/${productoId}`).then((r) => r.data),

  listarEntradas: (params: {
    fecha_inicio: string
    fecha_fin: string
    skip?: number
    limit?: number
  }) =>
    api
      .get<MovimientoEntradaReporte[]>('/inventario/movimientos/entradas', {
        params,
      })
      .then((r) => r.data),

  listarEntradasDetalle: (params: {
    fecha_inicio: string
    fecha_fin: string
    skip?: number
    limit?: number
  }) =>
    api
      .get<MovimientoEntradaDetalle[]>('/inventario/movimientos/entradas/detalle', {
        params,
      })
      .then((r) => r.data),

  listarReporteMovimientos: (params: {
    fecha_inicio: string
    fecha_fin: string
    producto_id?: number
    skip?: number
    limit?: number
  }) =>
    api
      .get<MovimientoInventarioReporteDetalle[]>('/inventario/movimientos/reporte', {
        params,
      })
      .then((r) => r.data),

  corregirMovimiento: (movimientoId: number, data: { nuevo_producto_id: number; nueva_cantidad: number }) =>
    api.patch<MovimientoInventario>(`/inventario/movimientos/${movimientoId}/corregir`, data).then((r) => r.data),

  registrarMovimiento: (data: MovimientoInventarioCreate) =>
    api.post<MovimientoInventario>('/inventario/movimientos', data).then((r) => r.data),
}

File: app/services/inventario_service.py
"""
Servicio de inventario - movimientos y actualización de stock.
"""
from app.core.exceptions import NotFoundError, ValidationError
from app.models import Inventario, MovimientoInventario
from app.repositories.inventario_repository import InventarioRepository
from app.repositories.movimiento_inventario_repository import MovimientoInventarioRepository


class InventarioService:
    def __init__(self, db):
        self.db = db
        self.repo = InventarioRepository(db)
        self.mov_repo = MovimientoInventarioRepository(db)

    def registrar_movimiento(
        self,
        producto_id: int,
        tipo: str,
        cantidad: int,
        motivo: str | None = None,
        referencia_tipo: str | None = None,
        referencia_id: int | None = None,
        usuario_id: int | None = None,
        commit: bool = True,
    ) -> MovimientoInventario:
        if tipo not in ("entrada", "salida", "ajuste"):
            raise ValidationError("Tipo de movimiento inválido")
        if cantidad == 0:
            raise ValidationError("La cantidad no puede ser cero")

        inv = self.repo.get_by_producto(producto_id)
        if not inv:
            inv = Inventario(producto_id=producto_id, stock_actual=0, stock_minimo=0)
            self.db.add(inv)
            self.db.flush()

        stock_anterior = inv.stock_actual
        if tipo in ("entrada", "ajuste"):
            stock_nuevo = stock_anterior + cantidad  # cantidad puede ser negativa
        else:
            stock_nuevo = stock_anterior - cantidad

        mov = MovimientoInventario(
            producto_id=producto_id,
            tipo=tipo,
            cantidad=cantidad,
            stock_anterior=stock_anterior,
            stock_nuevo=stock_nuevo,
            motivo=motivo,
            referencia_tipo=referencia_tipo,
            referencia_id=referencia_id,
            usuario_id=usuario_id,
        )
        self.db.add(mov)
        inv.stock_actual = stock_nuevo
        if commit:
            self.db.commit()
            self.db.refresh(mov)
        return mov

    def actualizar_stock(self, producto_id: int, nuevo_stock: int) -> Inventario:
        inv = self.repo.get_by_producto(producto_id)
        if not inv:
            raise NotFoundError("Inventario del producto no encontrado")
        inv.stock_actual = nuevo_stock
        self.db.commit()
        self.db.refresh(inv)
        return inv

    def corregir_movimiento(
        self,
        movimiento_id: int,
        nuevo_producto_id: int,
        nueva_cantidad: int,
        usuario_id: int | None = None,
    ) -> tuple[MovimientoInventario, MovimientoInventario]:
        """
        Corrige un movimiento de entrada: crea salida para revertir el original
        y entrada para el producto/cantidad corregido.
        """
        mov = self.mov_repo.get_by_id(movimiento_id)
        if not mov:
            raise NotFoundError("Movimiento no encontrado")
        if mov.tipo != "entrada":
            raise ValidationError("Solo se pueden corregir movimientos de entrada")

        # Revertir salida del producto original
        self.registrar_movimiento(
            producto_id=mov.producto_id,
            tipo="salida",
            cantidad=mov.cantidad,
            motivo="Corrección de entrada",
            usuario_id=usuario_id,
            commit=False,
        )

        # Nueva entrada con producto/cantidad corregido
        self.db.flush()
        nuevo_mov = self.registrar_movimiento(
            producto_id=nuevo_producto_id,
            tipo="entrada",
            cantidad=nueva_cantidad,
            motivo="Corrección de entrada",
            usuario_id=usuario_id,
            commit=True,
        )

        # Obtener el movimiento de salida creado
        salida = (
            self.db.query(MovimientoInventario)
            .filter(
                MovimientoInventario.producto_id == mov.producto_id,
                MovimientoInventario.tipo == "salida",
            )
            .order_by(MovimientoInventario.id.desc())
            .first()
        )
        return (salida, nuevo_mov)
